src/receive_imf.rs

```diff
@@ -1018,12 +1018,7 @@ async fn add_parts(
             }
         }
 
-        state = if seen
-            || fetching_existing_messages
-            || is_mdn
-            || is_reaction
-            || chat_id_blocked == Blocked::Yes
-        {
+        state = if seen || fetching_existing_messages || is_mdn || chat_id_blocked == Blocked::Yes {
```

Comment:
Ok, this is a bug which requires a separate test it seems.

But again, making reactions `InNoticed` to only make `marknoticed_chat()` faster doesn't look conceptually correct. Reactions are still a user-noticeable info, so there's smth to "notice". And wrt the bug you found, rather we shouldn't emit `IncomingMsg` for hidden messages. But not sure, need to try and see if the tests doesn't break at least.

Reply:
A PR extending the existing tests to catch this bug: #6257

Reply:
Not necessarily, `get_fresh_msg_cnt()` includes the `and m.hidden=0` SQL condition, so if we need, we can make incoming reactions hidden InFresh messages. As for "is shown in notifications", now this is true for incoming reactions, that's also why i think they should be InFresh. If we add them as InNoticed and show notifications for them, that would be different from incoming messages. So, my suggestion is:
- Add incoming reactions as as hidden InFresh messages.
- Add other incoming hidden messages as InSeen. But looking at `receive_imf.rs` i don't see such messages even exist. But then i don't understand why we have `and m.hidden=0` checks for incoming messages in SQL everywhere. So this needs investigation.
- Add a db migration for existing hidden InFresh messages. At least just in case.

Reply:
From a code perspective, we can make them InFresh, it requires one more special case (in order not to send an IncomingMsg), but that doesn't make any difference, really.

What does make a difference is whether we introduce 1ms of delay (1ms on quite a new phone, that is - older phones may easily need 3 times as long). Doing this just once isn't noticeable, but if we do it often (and without necessarily needing to) then we will end up with slow code (and markseen is called every time a chat is opened). Measuring this isn't hard, you can also do it - if you find a way that is just as fast, I'm fine with that, too. You may need to measure on a phone, not on Desktop, because on a fast PC the difference may not be noticeable.

> Add a db migration for existing hidden InFresh messages. At least just in case.

We shouldn't do database migrations "just in case", since there is always some chance that they create problems, and we don't have any indication that it will solve problems.

Reply:
> From a code perspective, we can make them InFresh, it requires one more special case (in order not to send an IncomingMsg)

Done in https://github.com/deltachat/deltachat-core-rust/pull/6213/commits/110f0d3c2e27cf37fd20d839a970c91ee517c9d6, doesn't look complex.

> What does make a difference is whether we introduce 1ms of delay

I'm going to profile that, but anyway, there's a new SQL statement selecting hidden messages (actually reactions), so i'm afraid it will always add some delay (NB: `message::markseen_msgs()` is a no-op if `msg_ids` is empty, so it's not a source of the problem). In https://github.com/deltachat/deltachat-core-rust/pull/6213#discussion_r1855222201 you suggested a solution:
> This can be fixed by writing `WHERE state=?` here and replacing `MessageState::InFresh, MessageState::InSeen` with `MessageState::InNoticed` below.

but i don't understand how it works -- just because you simplify the SQL statement (which is unlikely) or because you don't select other unnecessary messages and don't pass them to `message::markseen_msgs()`? But there should be no other hidden `InFresh` messages currently (NB: Only looking at current `receive_imf.rs`, i'm not sure they were never created in the past).

> We shouldn't do database migrations "just in case", since there is always some chance that they create problems

See "NB" above -- if hidden `InFresh` messages were created in the past, they will be selected and passed to `message::markseen_msgs()` which we should avoid. So, probably this db migration will be a no-op, but as we can't be sure (we can't inspect all the previously released versions, really), better to add it.

Reply:
> but i don't understand how it works -- just because you simplify the SQL statement (which is unlikely) or because you don't select other unnecessary messages and don't pass them to `message::markseen_msgs()`?

I don't know, my guess would be that it's because the simplification in the SQL statement enabled some optimization.

Reply:
Maybe it works because we have `msgs_index7 ON msgs (state, hidden, chat_id)`. If so, i should replace `state>=? AND state<?` with an `OR` condition containing just `InFresh` and `InNoticed`.